### PR TITLE
feat: add missing Type type guards

### LINKED ISSUE #3: 🚀 Feature: Fill in remaining tsutils functions (that don't exist on TypeScript)
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/ts-api-tools/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

I'm filing this as a catch-all issue for anybody interested in contributing! 👋 

[`tsutils`](https://github.com/ajafff/tsutils) has a lot of functions in it. Many of them are now available as part of this package. But some of them aren't. Welcoming contributions to add functions to this package that exist in `tsutils` but not yet in `ts-api-tools`.

The one exception to this is type guards such as `isBinaryExpression` that now exist on TypeScript itself. `tsutils` contains a lot of those functions for backwards compatibility with old `tsutils`+`typescript` versions - but this package doesn't need to.

### Additional Info

Any new file whose contents are essentially a copy & paste of a `tsutils` file should have the same licensing header as other source files of that nature:

```ts
// Code largely based on https://github.com/ajafff/tsutils
// Original license: https://github.com/ajafff/tsutils/blob/26b195358ec36d59f00333115aa3ffd9611ca78b/LICENSE
```

### PR DESCRIPTION
**Pull request by RebeccaStevens**
## PR Checklist

- [x] Addresses an existing open issue: fixes #3
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/ts-api-utils/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

Adds any missing type type guards from `tsutils` and adds new type type guards for other types.

### CODE REVIEW

File: src/types/typeGuards/objects.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+export function isTypeReference(type: ts.Type): type is ts.TypeReference {
+	return isObjectType(type) && isObjectFlagSet(type, ts.ObjectFlags.Reference);
+}
+
+/*
+ * No TypeScript defined type exists for types with these object flags.
+ */
+
+// export function isAnonymousType(type: ts.Type): boolean { return isObjectType(type) && isObjectFlagSet(type, ts.ObjectFlags.Anonymous); }
```
Comment (RebeccaStevens) [inline src/types/typeGuards/objects.ts:51]: Should this and the following function be uncommented?
They aren't technically type guards but they perform the same type of functionality.

File: src/types/typeGuards/simple.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	type: ts.Type
+): type is ts.UniqueESSymbolType {
+	return isTypeFlagSet(type, ts.TypeFlags.UniqueESSymbol);
+}
+
+/*
+ * No TypeScript defined type exists for types with these type flags.
+ */
+
+// export function isAnyType(type: ts.Type): boolean { return isTypeFlagSet(type, ts.TypeFlags.Any); }
```
Comment (RebeccaStevens) [inline src/types/typeGuards/simple.ts:101]: More function that could also be exported.

File: src/types/typeGuards/objects.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+export function isEvolvingArrayType(
+	type: ts.Type
+): type is ts.EvolvingArrayType {
+	return (
+		isObjectType(type) && isObjectFlagSet(type, ts.ObjectFlags.EvolvingArray)
+	);
+}
+
+export function isInterfaceType(type: ts.Type): type is ts.InterfaceType {
```
Comment (RebeccaStevens) [inline src/types/typeGuards/objects.ts:15]: This type is really confusing me. We need to look into this more.

The comment in the TypeScript type declarations file seems to imply that `ObjectFlags.Class` or `ObjectFlags.Interface` should be set on any instance of this type.

However, if this type supposed to include types declared by an interface, then the fields on `InterfaceType` don't line up with the runtime data; so `InterfaceType` must not include such types (hence [this test](https://github.com/RebeccaStevens/ts-api-utils/blob/db5958611b42655486e47b4ef01c34f32695f747/src/types/typeGuards/objects.test.ts#LL22C3-L22C45).
Classes and instances of classes do have the right runtime data though. They also have the `ObjectFlags.Class` object flag; so everything is right for them.

But there are types that they are an `InterfaceType` but though have either `ObjectFlags.Class` or `ObjectFlags.Interface` set. For example, `TupleType` extends `GenericType` which extends `InterfaceType`.

**Comment by JoshuaKGoldberg** [comment]
Marking as blocked on #51

**Comment by RebeccaStevens** [comment]
This one shouldn't be marked as blocked by #51
These functions aren't defined in wither TypeScript's public or private API.

File: src/types/typeGuards/objects.ts (hunk 1 above)
Comment (RebeccaStevens) [reply src/types/typeGuards/objects.ts:51]: These functions are for each value in the `ts.ObjectFlags` enum.

File: src/flags.ts [hunk 4]
```diff
@@ -3,31 +3,42 @@
```
[the remark this answers is not in the document]
Comment (RebeccaStevens) [reply src/flags.ts:3]: Essentially the way I've been making these updates is by querying the `typescript.d.ts` file for types (via `CTRL+F` and `CTRL+D`), copying all the matching text out, and then turning them into functions. I haven't been using the original `tsutils` code at all, other than to check if I missed a function or two.

Here's a quick video of how I start these sorts of PRs:

https://user-images.githubusercontent.com/7224206/218638312-9b46cb98-3c38-4098-8585-17f745399a5e.mp4

So, in that regard, the old license shouldn't be needed as all of the code is new.

File: src/types/typeGuards/objects.test.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	const { sourceFile, typeChecker } =
+		createSourceFileAndTypeChecker(sourceText);
+	const node = sourceFile.statements.at(-1)!;
+
+	const type = typeChecker.getTypeAtLocation(node);
+	if ((type as { intrinsicName?: string }).intrinsicName === "error") {
+		throw new Error("test case error");
+	}
+	return type;
+}
```
[the remark this answers is not in the document]
Comment (RebeccaStevens) [reply src/types/typeGuards/objects.test.ts:16]: Sound like a good idea.

File: src/flags.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/flags.ts:3]: Love it!

File: src/types/typeGuards/objects.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply src/types/typeGuards/objects.ts:51]: Yeah I'd say let's remove them then.

File: src/types/typeGuards/objects.ts (hunk 3 above)
Comment (RebeccaStevens) [reply src/types/typeGuards/objects.ts]: I've remove this function for now. We can look into it again if future.

**Comment by RebeccaStevens** [comment]
I've removed those comment out function and the `isInterfaceType` function.
We can look into those in future.
PR should be unblocked now.

File: src/flags.ts [hunk 6]
```diff
@@ -3,31 +3,42 @@
 
 import * as ts from "typescript";
 
-function isFlagSet(obj: { flags: number }, flag: number): boolean {
-	return (obj.flags & flag) !== 0;
+function isFlagSet(allFlags: number, ...flags: number[]): boolean {
+	return (allFlags & flags.reduce((carry, flag) => carry | flag)) !== 0;
+}
```
[the remark this answers is not in the document]
Comment (RebeccaStevens) [reply src/flags.ts:8]: I'll revert those changes.

**Pull request merged**